runtime/compiler/z/codegen/J9S390JNILinkage.cpp

```diff
@@ -0,0 +1,964 @@
+                                                                      refPoolSnippetLabel, false, true);
+   gcPoint->setNeedsGCMap(0);
+
+   TR::SymbolReference * collapseSymRef = cg()->symRefTab()->findOrCreateRuntimeHelper(TR_S390collapseJNIReferenceFrame,
+                                                                                       false, false, false);
+
+   cg()->addSnippet(new (trHeapMemory()) TR::S390HelperCallSnippet(cg(), callNode,
+      refPoolSnippetLabel, collapseSymRef, refPoolRestartLabel));
+   generateS390LabelInstruction(cg(), TR::InstOpCode::LABEL, callNode, refPoolRestartLabel);
+   }
+
+void
+TR::J9S390JNILinkage::unwrapReturnValue(TR::Node* callNode,
+                                        TR::Register* javaReturnRegister,
+                                        TR::RegisterDependencyConditions* deps)
+   {
+   TR::LabelSymbol * tempLabel = generateLabelSymbol(cg());
+   TR::Instruction* start = generateS390CompareAndBranchInstruction(cg(), TR::InstOpCode::getCmpOpCode(),
+                                                                    callNode, javaReturnRegister,
+                                                                    (signed char)0, TR::InstOpCode::COND_BE, tempLabel);
+
+   start->setStartInternalControlFlow();
+   generateRXInstruction(cg(), TR::InstOpCode::getLoadOpCode(), callNode, javaReturnRegister,
+              generateS390MemoryReference(javaReturnRegister, 0, cg()));
+
+   int32_t regPos = deps->searchPostConditionRegisterPos(javaReturnRegister);
+   TR::RealRegister::RegNum realReg = deps->getPostConditions()->getRegisterDependency(regPos)->getRealRegister();
+   TR::RegisterDependencyConditions * postDeps = new (trHeapMemory()) TR::RegisterDependencyConditions(0, 1, cg());
+   postDeps->addPostCondition(javaReturnRegister, realReg);
+
+   TR::Instruction* end =generateS390LabelInstruction(cg(), TR::InstOpCode::LABEL, callNode, tempLabel);
+   end->setEndInternalControlFlow();
+   end->setDependencyConditions(postDeps);
+   }
+
+extern "C" void fast_jitPreJNICallOffloadCheck();
+extern "C" void fast_jitPostJNICallOffloadCheck();
+
+void
+TR::J9S390JNILinkage::callIFAOffloadCheck(TR::RegisterDependencyConditions* deps,
+                                          bool isJNICallOutFrame,
+                                          bool isPreJNIHelper)
+   {
+   TR_RuntimeHelper helperIndex = isPreJNIHelper ? TR_S390jitPreJNICallOffloadCheck : TR_S390jitPostJNICallOffloadCheck;
+   const char* symName  = isPreJNIHelper ? "fast_jitPreJNICallOffloadCheck" : "fast_jitPostJNICallOffloadCheck";
+   void* targetAddr     = isPreJNIHelper ? (void*)fast_jitPreJNICallOffloadCheck : (void*)fast_jitPostJNICallOffloadCheck;
+
+   TR::SymbolReference * offloadSymRef = cg()->symRefTab()->findOrCreateRuntimeHelper(helperIndex, false, false, false);
+   offloadSymRef->getSymbol()->setName(symName);
+   offloadSymRef->getSymbol()->castToMethodSymbol()->setMethodAddress(targetAddr);
```

Comment:
I don't think we need this code as this is already handled when the helper is created here:

https://github.com/eclipse/omr/blob/70ff3c7d2a7fa2f85f380dadf23e5acda65a45bc/compiler/compile/OMRSymbolReferenceTable.cpp#L714-L735

Are we able to remove this and simply pass in the `TR_RuntimeHelper` index instead of `bool isPreJNIHelper`? We can also remove the above two `extern "C"` definitions and simplify this function greatly. In fact it may become so small as to not even warrant an existence?

Reply:
Removed the `extern` statements and used `TR_RuntimeHelper` indices instead.